File: camera-ui-wasm-motion/src/sensor.ts
import { instantiate } from '@assemblyscript/loader';
import { MotionDetectorSensor } from '@camera.ui/sdk';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import { DEFAULT_BLUR_RADIUS, DEFAULT_DILATION_SIZE, DEFAULT_MOTION_AREA, DEFAULT_THRESHOLD } from './defaults.js';

import type { ASUtil } from '@assemblyscript/loader';
import type { JsonSchema, MotionResult, VideoFrameData } from '@camera.ui/sdk';
import type { instantiate as instantiateWasm } from './wasm/build/detector.js';

type WASMModule = ASUtil & Record<string, unknown> & Awaited<ReturnType<typeof instantiateWasm>>;

export interface WASMMotionStorageValues {
  area: number;
  threshold: number;
  blurRadius: number;
  dilationSize: number;
}

export class WASMMotionSensor extends MotionDetectorSensor<WASMMotionStorageValues> {
  private wasmExports?: WASMModule;
  private frameView?: Uint8Array;
  private boxesView?: Int32Array;
  private lastWidth = 0;
  private lastHeight = 0;

  readonly isAvailable = true;

  constructor(name = 'WASM Motion') {
    super(name);
  }

  get storageSchema(): JsonSchema[] {
    return [
      {
        type: 'number',
        key: 'area',
        title: 'Area',
        description: 'Smallest region size that counts as motion.',
        store: true,
        defaultValue: DEFAULT_MOTION_AREA,
        minimum: 10,
        maximum: 1000,
        step: 1,
        required: true,
      },
      {
        type: 'number',
        key: 'threshold',
        title: 'Threshold',
        description: 'Brightness change needed to mark a pixel as changed.',
        store: true,
        defaultValue: DEFAULT_THRESHOLD,
        minimum: 1,
        maximum: 255,
        step: 1,
        required: true,
      },
      {
        type: 'number',
        key: 'blurRadius',
        title: 'Blur Radius',
        description: 'Smoothing applied before detection to reduce noise.',
        store: true,
        defaultValue: DEFAULT_BLUR_RADIUS,
        minimum: 1,
        maximum: 21,
        step: 1,
        required: true,
      },
      {
        type: 'number',
        key: 'dilationSize',
        title: 'Dilation Size',
        description: 'Merges nearby changed pixels into one region.',
        store: true,
        defaultValue: DEFAULT_DILATION_SIZE,
        minimum: 1,
        maximum: 21,
        step: 1,
        required: true,
      },
      {
        type: 'button',
        key: 'default',
        title: 'Default Settings',
        description: 'Restore the default detection values.',
        color: 'danger',
        onSet: async () => {
          await this.resetToDefaults();
        },
      },
    ];
  }

  async detectMotion(frame: VideoFrameData): Promise<MotionResult> {
    await this.ensureWASMInitialized(frame.width, frame.height);

    if (!this.wasmExports) {
      return { detected: false, detections: [] };
    }

    if (!this.frameView || !this.boxesView) {
      return { detected: false, detections: [] };
    }

    const config = this.getConfig();

    const { detectMotion } = this.wasmExports;

    const frameData = frame.data instanceof Uint8Array ? frame.data : new Uint8Array(frame.data);
    this.frameView.set(frameData.subarray(0, this.frameView.length));

    const numBoxes = detectMotion(config.threshold, config.blurRadius, config.dilationSize, config.area);

    const detections: MotionResult['detections'] = [];

    for (let i = 0; i < numBoxes; i++) {
      const x = this.boxesView[i * 4];
      const y = this.boxesView[i * 4 + 1];
      const w = this.boxesView[i * 4 + 2];
      const h = this.boxesView[i * 4 + 3];

      detections.push({
        label: 'motion',
        confidence: 1,
        box: {
          x: x / frame.width,
          y: y / frame.height,
          width: w / frame.width,
          height: h / frame.height,
        },
      });
    }

    return {
      detected: detections.length > 0,
      detections,
    };
  }

  resetState(): void {
    this.wasmExports?.reset();
  }

  private getConfig(): WASMMotionStorageValues {
    return {
      area: this.storage.values.area,
      threshold: this.storage.values.threshold,
      blurRadius: this.storage.values.blurRadius,
      dilationSize: this.storage.values.dilationSize,
    };
  }

  private async ensureWASMInitialized(width: number, height: number): Promise<void> {
    if (this.wasmExports && this.lastWidth === width && this.lastHeight === height) {
      return;
    }

    if (!this.wasmExports) {
      const wasmModule = await instantiate(readFileSync(resolve(__dirname, 'detector.wasm')), {
        console: {
          log: () => {},
        },
        env: {
          abort: () => {},
        },
      });

      this.wasmExports = wasmModule.exports as WASMModule;
    }

    this.wasmExports.initialize(width, height);

    const exports = this.wasmExports;
    const buffer = exports.memory.buffer as ArrayBuffer;
    this.frameView = new Uint8Array(buffer, exports.getFramePtr(), width * height);
    this.boxesView = new Int32Array(buffer, exports.getBoxesPtr(), exports.getMaxBoxes() * 4);

    this.lastWidth = width;
    this.lastHeight = height;
  }

  private async resetToDefaults(): Promise<void> {
    if (this.storage) {
      await this.storage.setValue('area', DEFAULT_MOTION_AREA);
      await this.storage.setValue('threshold', DEFAULT_THRESHOLD);
      await this.storage.setValue('blurRadius', DEFAULT_BLUR_RADIUS);
      await this.storage.setValue('dilationSize', DEFAULT_DILATION_SIZE);
    }
  }
}

File: camera-ui-opencl/src/sensor.py
from __future__ import annotations

import asyncio
from concurrent.futures import ThreadPoolExecutor
from contextlib import suppress
from typing import Any, TypedDict

import numpy as np
from camera_ui_sdk import (
    CameraDevice,
    JsonSchema,
    MotionDetectorSensor,
    MotionResult,
    VideoFrameData,
)

from detector_defaults import (
    DEFAULT_ALPHA,
    DEFAULT_AREA,
    DEFAULT_BLUR,
    DEFAULT_DILT,
    DEFAULT_THRESHOLD,
)
from opencl_detector import OpenCLMotionDetector, create_program


class OpenCLStorageValues(TypedDict):
    area: int
    threshold: float
    blur: int
    dilation: int


class OpenCLMotionSensor(MotionDetectorSensor[OpenCLStorageValues]):
    def __init__(self, camera_device: CameraDevice, name: str = "OpenCL Motion") -> None:
        super().__init__(name)

        self._camera_device = camera_device
        self._opencl_ctx: tuple[Any, Any, Any] | None = None
        self._opencl_detector: OpenCLMotionDetector | None = None
        self._executor: ThreadPoolExecutor | None = None
        self._is_available = False

        with suppress(Exception):
            self._opencl_ctx = create_program()
            self._is_available = True

    @property
    def storage_schema(self) -> list[JsonSchema]:
        return [
            {
                "type": "number",
                "key": "area",
                "title": "Area",
                "description": "Minimum size of detected motion (pixels)",
                "store": True,
                "defaultValue": DEFAULT_AREA,
                "minimum": 10,
                "maximum": 1000,
                "step": 1,
                "required": True,
            },
            {
                "type": "number",
                "key": "threshold",
                "title": "Threshold",
                "description": "Sensitivity of motion detection (0-1, higher = less sensitive)",
                "store": True,
                "defaultValue": DEFAULT_THRESHOLD,
                "minimum": 0,
                "maximum": 1,
                "step": 0.01,
                "required": True,
            },
            {
                "type": "number",
                "key": "blur",
                "title": "Blur",
                "description": "Gaussian blur radius to reduce noise",
                "store": True,
                "defaultValue": DEFAULT_BLUR,
                "minimum": 1,
                "maximum": 21,
                "step": 2,
                "required": True,
            },
            {
                "type": "number",
                "key": "dilation",
                "title": "Dilation",
                "description": "Expansion of detected motion areas",
                "store": True,
                "defaultValue": DEFAULT_DILT,
                "minimum": 1,
                "maximum": 21,
                "step": 2,
                "required": True,
            },
            {
                "type": "button",
                "key": "reset_defaults",
                "title": "Reset to Defaults",
                "description": "Reset motion detection settings to default",
                "color": "danger",
                "onSet": self._reset_to_defaults,
            },
        ]

    @property
    def isAvailable(self) -> bool:
        return self._is_available

    async def detectMotion(self, frame: VideoFrameData) -> MotionResult:
        if not self._opencl_ctx:
            return {"detected": False, "detections": []}

        current_frame: np.ndarray[Any, Any] = np.frombuffer(frame["data"], dtype=np.uint8).reshape(
            (frame["height"], frame["width"])
        )

        width = frame["width"]
        height = frame["height"]

        if self._executor is None:
            self._executor = ThreadPoolExecutor(max_workers=2, thread_name_prefix="OpenCL")

        loop = asyncio.get_event_loop()

        if self._opencl_detector is None:
            self._opencl_detector = await loop.run_in_executor(
                self._executor,
                OpenCLMotionDetector,
                self._opencl_ctx,
                width,
                height,
                self.storage.values["blur"],
                self._camera_device.logger,
            )

        detections = await loop.run_in_executor(
            self._executor,
            self._opencl_detector.process_frame,
            current_frame,
            self.storage.values["threshold"],
            self.storage.values["dilation"],
            self.storage.values["area"],
            DEFAULT_ALPHA,
        )

        return {
            "detected": len(detections) > 0,
            "detections": [
                {
                    "label": "motion",
                    "confidence": 1.0,
                    "box": {
                        "x": det[0] / width,
                        "y": det[1] / height,
                        "width": (det[2] - det[0]) / width,
                        "height": (det[3] - det[1]) / height,
                    },
                }
                for det in detections
            ],
        }

    def resetState(self) -> None:
        self._opencl_detector = None
        if self._executor:
            self._executor.shutdown(wait=False)
            self._executor = None

    def on_deassigned(self) -> None:
        self.resetState()

    async def _reset_to_defaults(self) -> None:
        if self.storage:
            await self.storage.setValue("area", DEFAULT_AREA)
            await self.storage.setValue("threshold", DEFAULT_THRESHOLD)
            await self.storage.setValue("blur", DEFAULT_BLUR)
            await self.storage.setValue("dilation", DEFAULT_DILT)
